File: module/ImaginaryCraft/src/main/java/architecture/imaginarycraft/init/IcItems.java
package architecture.imaginarycraft.init;

import architecture.goldenboughs_lib.common.item.FoodItem;
import architecture.goldenboughs_lib.util.FoodPropertiesBuilder;
import architecture.imaginarycraft.datagen.i18n.IcZhCn;
import architecture.imaginarycraft.util.IcUtil;
import net.minecraft.world.item.Item;
import net.neoforged.neoforge.registries.DeferredItem;
import net.neoforged.neoforge.registries.DeferredRegister;

import java.util.function.Function;

public final class IcItems {
    public static final DeferredRegister.Items REGISTRY = DeferredRegister.Items.createItems(IcUtil.ID);

    /**
     * 罐装脑啡肽
     */
    public static final DeferredItem<FoodItem> CANNED_ENKEPHALIN = registerFood(
            "canned_enkephalin", "罐装脑啡肽", false,
            new FoodPropertiesBuilder()
                    .eatSeconds(84)
                    .alwaysEdible()
    );

    private IcItems() {
    }

    //region 食物/饮品注册

    /**
     * 注册食物/饮品（使用默认 func）。
     */
    private static DeferredItem<FoodItem> registerFood(
            String name, String nameZh, boolean isEat,
            Item.Properties properties,
            FoodPropertiesBuilder foodPropertiesBuilder
    ) {
        return registerFood(name, nameZh, isEat, properties, foodPropertiesBuilder,
                props -> new FoodItem(props, foodPropertiesBuilder.build(), isEat));
    }

    /**
     * 注册食物/饮品（使用默认 {@link Item.Properties}）。
     */
    private static DeferredItem<FoodItem> registerFood(
            String name, String nameZh, boolean isEat,
            FoodPropertiesBuilder foodPropertiesBuilder
    ) {
        return registerFood(name, nameZh, isEat, new Item.Properties(), foodPropertiesBuilder);
    }

    /**
     * 注册食物/饮品（完整参数）。
     */
    private static DeferredItem<FoodItem> registerFood(
            String name, String nameZh, boolean isEat,
            Item.Properties properties,
            FoodPropertiesBuilder foodPropertiesBuilder,
            Function<Item.Properties, FoodItem> func
    ) {
        return register(name, nameZh, func, properties);
    }

    //endregion

    //region 通用物品注册

    private static <I extends Item> DeferredItem<I> register(
            String name, String nameZh,
            Function<Item.Properties, I> func,
            Item.Properties properties
    ) {
        var registerItem = REGISTRY.registerItem(name, func, properties.stacksTo(1));
        IcZhCn.addI18nItemText(nameZh, registerItem);
        return registerItem;
    }

    //endregion
}

File: module/ImaginaryCraft/src/main/kotlin/architecture/imaginarycraft/init/IcItems.kt
package architecture.imaginarycraft.init

import architecture.goldenboughs_lib.common.item.FoodItem
import architecture.goldenboughs_lib.util.FoodPropertiesBuilder
import architecture.imaginarycraft.core.ImaginaryCraftConstants
import architecture.imaginarycraft.datagen.i18n.LcZhCn
import net.minecraft.world.item.Item
import net.neoforged.neoforge.registries.DeferredItem
import net.neoforged.neoforge.registries.DeferredRegister
import java.util.function.Function

object IcItems {
	@JvmField
	val REGISTRY: DeferredRegister.Items = DeferredRegister.Items.createItems(ImaginaryCraftConstants.ID)

	@JvmField
	val CANNED_ENKEPHALIN = registerFood(
		"canned_enkephalin", "罐装脑啡肽", false,
		foodPropertiesBuilder = FoodPropertiesBuilder()
			.eatSeconds(84)
			.alwaysEdible()
	)

	//	DRINK
	private fun registerFood(
		name: String,
		nameZh: String,
		isEat: Boolean = true,
		properties: Item.Properties = Item.Properties(),
		foodPropertiesBuilder: FoodPropertiesBuilder,
		func: Function<Item.Properties, FoodItem> = { FoodItem(it, foodPropertiesBuilder.build(), isEat) }
	): DeferredItem<FoodItem> {
		return register(name, nameZh, func, properties)
	}

	private fun <I : Item> register(
		name: String,
		nameZh: String,
		func: Function<Item.Properties, I>,
		properties: Item.Properties
	): DeferredItem<I> {
		val registerItem = REGISTRY.registerItem(
			name, func, properties
				.stacksTo(1)
		)
		LcZhCn.addI18nItemText(nameZh, registerItem)
		return registerItem
	}
}
